### DISCUSSION TITLE
Enabling plugin via command line works, but also does not work. How to enable?

### DISCUSSION THREAD

**Discussion by bencresty** [craftcms/cms#14173] [Q&A]
Hi,
When enabling a plugin that didn't change, but is running locally on another system now, it is not enabling in the cms settings when hitting 'enable'.

When running craft plugin/enable interactively and chosing these plugins the command says 'enabled {plugin} successfully', but when running the craft plugin/enable command again it's still in the list of disabled plugins and it's also still disabled according to the cms.

It's not giving any clue on what might be wrong and the plugin was working fine before and didn't change since. Only craft has been updated once in a while and the project was moved to another system.

I already checked if there was an issue with usages of casing in the classes of my plugins, but everything seems okay. Which doesn't surprise me as these were working fine for years in many Craft versions and still are to this date on the servers these are running on publicly. 

I also ran the composer dump-autoload again, whithout any issue.

It would be nice to have some form of clue on where to look. So any advise would be welcome.

Currently using craft 3.9.10, but it has been like this for a while now, also with previous versions.

Any tips on where to look?

**Comment by brandonkelly** [comment]
Do you have the `disabledPlugins` config setting set in `config/general.php`, or via the `CRAFT_DISABLED_PLUGINS` environment variable?

**Reply by bencresty** [reply]
Thanks for your reply. No. There's no such setting defined. 

![image](https://github.com/craftcms/cms/assets/69322733/bc889f2c-1cfa-4615-ab03-70dae0ea68f1)

**Reply by brandonkelly** [reply]
And if you go to **Utilities** → **PHP Info** and search for `CRAFT_DISABLED_PLUGINS`, no results show up?

**Reply by bencresty** [reply]
Indeed. Nothing turns up

**Reply by brandonkelly** [reply]
Hm, in that case I’m not really sure. Tested and `plugin/disable` and `plugin/enable` are both working as expected for me.

Are you able to send your database and `composer.json` + `composer.lock` files to support@craftcms.com?

**Reply by bencresty** [reply]
You know by earlier conversations that I may never send these files. If there's something I could check here though I'd be happy to know

**Reply by brandonkelly** [reply]
Plugin enabled/disabled status is only stored in the project config. Are other PC-only settings getting saved properly? (Email settings, User settings, System Name, System Status, and Time Zone)

**Reply by bencresty** [reply]
Thanks. That's a good test indeed. When I get back I'll see if the other settings get saved

**Reply by bencresty** [reply]
@brandonkelly I've just changed the 'System Name' and it was saved just fine (was still having the same name after reloading the settings page). 

Weird thing though: When enabling the 'feed-me' plugin in the project.yaml by changing 'enabled' from 'false' to 'true' this is recognised by the cms on the dev server as it states that something has changed. But when applying it doesn't even show feed-me in the whole plugin list as shown in the cms. So for whatever reason Craft seems to find it completely normal to enable a plugin that isn't even on the database. Not sure what it's doing there and why even the project.yaml could be different compared to the database. Honestly it remains error prone to have two 'sources of truth' (one in a yaml file and one in a database). I'll keep looking.

**Reply by brandonkelly** [reply]
> But when applying it doesn't even show feed-me in the whole plugin list as shown in the cms. So for whatever reason Craft seems to find it completely normal to enable a plugin that isn't even on the database.

Not sure I’m following. Is Feed Me actually Composer-installed?

**Reply by bencresty** [reply]
Well I would say it had to be like this once, but it was not in the composer.json here now, which is probably because I didn't need it after launch of the website years ago. Looks like the project.yaml never updated after removing it, but it's a long time ago so it's hard to tell.

For the record I'm the only one ever working on this site and I'm very precise and have the project yaml files checked into github, so if anything ever changed at least I must've seen this. But it's a long time ago since the last time I used this website locally. So I installed the production database on the dev server and this must exactly match the local yaml files, because everything is always deployed on the production server via git from the exact same files as I have locally.

Either way, I've just composer installed feed-me and went back to the cms, but now there's another issue; It tries to complete the installation, but trows the following error message now:
"Database Exception: SQLSTATE[42S21]: Column already exists: 1060 Duplicate column name 'singleton'
The SQL being executed was: ALTER TABLE feedme_feeds ADD singleton tinyint(1) NOT NULL DEFAULT FALSE AFTER sortOrder ..."

So it looks like the installation doesn't check first to see if the column exists.

I'm not too familiar with the database structure of Craft. Could I just remove the feedme_feeds table and try it again? Or is there more involed in this?

**Reply by bencresty** [reply]
BTW For the record, I would never remove a plugin like feed-me by hand using composer. I would always do it the right way via either the command line or via the 'UI'. So not sure how things happened and it's too long ago to check, but it surely wasn't a user error like that.

**Reply by bencresty** [reply]
BTW it's also impossible to 'Abort the update' it want to do, because if we click 'Abort Update' it goes back to this. So it doesn't really seem to abort. In this state the only thing possible to really abort seems to be to remove the database again and create it again from the backup. To have the same issue again later
![image](https://github.com/craftcms/cms/assets/69322733/a9809bea-f83f-40ce-b037-fd4590047b2a)

**Reply by bencresty** [reply]
And if we do click Finish here (after the abortion) we get this:
![image](https://github.com/craftcms/cms/assets/69322733/4afdda17-3709-4530-8b02-55a6cca17a93)

**Reply by bencresty** [reply]
Strangely enough:

I now removed the database and put the backup database (which is the current production database) up again on the dev-server, and now it DOES work. It doesn't show me the message to finish the installation and the feed-me plugin is just in the list on the cms plugins.
The only difference is that this time feed-me was installed with composer, installed via the craft command line and enabled via the craft command line BEFORE putting back the database.

So it looks like either Craft or the Feed-me is having issues installing when something was already installed, but something was out of sync. And when in that state it seems difficult to get out of that state if there's no database to restore to as I did now.

**Reply by bencresty** [reply]
Oh no. I was wrong. When I now click 'Enable' feed-me in the plugins list the whole rambam happens all over again. Yet again it is not able to finish the installation because it already finds these columns :(

So back to zero again :(

**Reply by bencresty** [reply]
And it gets even sillyer now;
When dropping the table it was complaining about it having already the columns and hitting 'Finish' again, now it complains that it cannot find the feedme_feeds table.... omg

![image](https://github.com/craftcms/cms/assets/69322733/379e8c06-18f6-46e5-9030-55e52fbd179d)

**Reply by brandonkelly** [reply]
> Well I would say it had to be like this once, but it was not in the composer.json here now, which is probably because I didn't need it after launch of the website years ago. Looks like the project.yaml never updated after removing it, but it's a long time ago so it's hard to tell.

Gotcha – so it must have been Composer-uninstalled before being uninstalled within Craft. Not really a “multiple sources of truth” issue; Craft just doesn’t go straight to assuming a plugin should be considered uninstalled if its files don’t exist. (Maybe it was an accidental Composer fluke or something). Plus, Craft doesn’t even know how to uninstall plugins that aren’t present, beyond simply removing their records from the `plugins` table and within the `plugins` array in the project config. Ideally you are having the plugin uninstall itself (via the “Uninstall” action, or `plugin/uninstall` command), propagating that change to other environments via `project.yaml`, and _then_ Composer-uninstalling it.

> BTW For the record, I would never remove a plugin like feed-me by hand using composer. I would always do it the right way via either the command line or via the 'UI'. So not sure how things happened and it's too long ago to check, but it surely wasn't a user error like that.

If you search through your Git history you should be able to find the commit where it was removed from `composer.json`. If `project.yaml` wasn’t removed at the same time, there’s the proof that it was only Composer-uninstalled. And if it was, then you could git-blame the `plugins.feed-me` config in `project.yaml` to see when it came back.

Regarding the new installation issues you’re facing, it sounds like we need to make the Feed Me installer quite a bit more defensive about preexisting tables.

Your best bet is probably to just `composer remove craftcms/feed-me`, and then run `php craft plugin/uninstall feed-me`. Then delete any `feedme_*` database tables that are left.

**Reply by bencresty** [reply]
I see you're defending Craft a little. Which makes complete sense since you're the creator. Without repeating what's already been said it's better to just agree to dissagree on the 2 sources of truth discussion as it's not helping either of us. I shouldn't have brought it up, eventhough it's hard for me not to in situations like this. I'll try harder next time. Please note though that this is in no way a lack of respect towards you, the team or the product though. Hope that's clear.

> Regarding the new installation issues you’re facing, it sounds like we need to make the Feed Me installer quite a bit more defensive about preexisting tables.

Thanks. That's very appreciated!

> Your best bet is probably to just `composer remove craftcms/feed-me`, and then run `php craft plugin/uninstall feed-me`. Then delete any `feedme_*` database tables that are left.

Good point. I'll try that. Thanks again

**Reply by brandonkelly** [reply]
Not defending Craft – just saying that plugins uninstalled with Composer first would have left artifacts in the database, even before Project Config.

**Reply by bencresty** [reply]
Which is an odd response after me telling you explicitely that I didn't do that. But leave it. That's why I wrote 'agree to dissagree'. There is an issue that needs to be solved now. I'm currently trying the uninstallation of feed-me route you advised. Which sounds good

**Reply by bencresty** [reply]
Yes, this did the trick. And after this I can now even enable and disable the plugin from within the UI, so that works too now. Weirdly enough I still cannot enable the local plugins, which was the original issue I started the tread for, but the feed-me one works now.

This is what I've done:
- your advise on how to remove and uninstall feed-me now (I needed to use the --force parameter as it wouldn't let me uninstall the disabled plugin otherwise)
- composer require craftcms/feed-me
- craft plugin/install feed-me
- craft plugin/enable feed-me

And than I just had the feed-me installed without needing to finish installation via the ui and the enable/disable on the feed-me work too now. 

Now it's time again to see if I can fix the original issue with the local plugins not being enabled and/or installed. 

I've just now tried to run just a 'composer install' to be sure everything in the composer.json is installed, but it says it doesn't have anything to install.  The path to the plugins however never changed and is a relative path to a folder right inside the project folder which always have been and is still there. So I wouldn't know a reason to why suddenly it doesn't work no more, but the plugins should be installed now.

**Reply by bencresty** [reply]
Alright, some weird things happening here now while trying to fix the issue with the local plugins not enabling in the UI:

I tried more or less the same approuch as with the feed-me that worked there here now:
- `craft plugin/uninstall sub-folder-name/plugin-name` --> shows 'uninstalled successfully'
- However, at this point the plugin according to the UI is still on 'Disabled', which is weird, as I would expect this to say 'Not ininstalled' now, just like another plugin in the list that's currently not installed. But either way I continued:

- `craft plugin/install sub-folder-name/plugin-name` --> strangely enough shows that the plugin-name handle is not recognised, while it was recognised right before this, so I try this (without the sub-folder-name)
- `craft plugin/install plugin-name` --> now it says 'installed succesfully' (beats me, why, because before it needed the sub-folder-name to recognise the handle?)
- No change in the UI at this point

- `craft plugin/enable plugin-name` --> now it says 'enabled successfully', but the UI of the CMS still shows no change (still shows 'disabled'. 

I'm sure it's me missing something, but to me this whole sequence doesn't make any sense to even remotely know where to look.

(BTW for the record again; I did a composer install to be sure, even a composer dump-auto-load, before all of this, although that shouldn't be needed at this point)

Conclusion for me at this point;
- The 'succesfully' message doesn't correspond to the findings when looking in the UI of the CMS. The UI of the CMS doesn't change a bit whatever I do and keeps saying these addons are disabled. Changing this to 'Enabled' from within the CMS itself doesn't work either, which was the issue it all started with.

Unfortunately I'm facing a tight deadline and a little bit out of ideas on what to try next honestly. I'm open to suggestions.

**Reply by bencresty** [reply]
Another thing to note: in the project.yaml these local plugins are already set to enabled: true. So these aren't disabled in the settings, and while they are installed according to the command line I wouldn't know why the CMS still thinks these are disabled (and don't use them on the pages)

**Reply by brandonkelly** [reply]
> `craft plugin/uninstall sub-folder-name/plugin-name` --> shows 'uninstalled successfully'

`sub-folder-name/plugin-name` is not a valid syntax for specifying a plugin. Only the `plugin-handle` is supported here.

If you have a plugin installed in a subfolder, you must pull it into the main Composer installation using a `path` repository in `composer.json`:

```json
"repositories": [
  {"type": "path", "url": "./sub-folder-name"}
]
```

and then run `composer require package/name`. Then you can install/uninstall/enable/disable it using its plugin handle, without any path leading up to it.

`plugin/uninstall` said the command was successful only because it first checks if a plugin is already installed with the exact handle you passed in, and if not, it will immediately abort and claim that it was successful:

https://github.com/craftcms/cms/blob/759edf01fecfe8efb0ccd1d372c123766fa4e660/src/services/Plugins.php#L586-L589

I suppose we could make that more defensive by ensuring that `$handle` is actually a valid plugin handle first, and throw an exception if not.

**Comment by bencresty** [comment]
> `sub-folder-name/plugin-name` is not a valid syntax for specifying a plugin. Only the `plugin-handle` is supported here.

It's been a while since I've worked with Craft, but again; this website worked fine before just as is. The only thing changed is that Craft was updated once in a while to new minor versions. And after that there never was any issue on the production server, still now there still isn't there. The problems now are only on the dev server.

> If you have a plugin installed in a subfolder, you must pull it into the main Composer installation using a `path` repository in `composer.json`:
> 
> ```json
> "repositories": [
>   {"type": "path", "url": "./sub-folder-name"}
> ]
> ```

Yes, that's exactly how it was built and worked all this time.

> and then run `composer require package/name`. Then you can install/uninstall/enable/disable it using its plugin handle, without any path leading up to it.

Also this I've already done long time ago, but for the sake of testing I did it once again. Didn't change anything to the composer.lock file, which shows it was already done.

> `plugin/uninstall` said the command was successful only because it first checks if a plugin is already installed with the exact handle you passed in, and if not, it will immediately abort and claim that it was successful:
> 
> https://github.com/craftcms/cms/blob/759edf01fecfe8efb0ccd1d372c123766fa4e660/src/services/Plugins.php#L586-L589
> 
> I suppose we could make that more defensive by ensuring that `$handle` is actually a valid plugin handle first, and throw an exception if not.

Yeah it would be expected that functions don't act like things worked great if in reality they couldn't even find the handle.


Anyway, it's still not working and stilll weird things are happening.

The plugins are installed fine with composer. No issue there and if composer says things were successful they were succesful, which also is proven by the fact that it didn't change anything to the lock file I would say.

When running a list to show the plugins' handles they are in the list just fine. (craft plugin/list). They show atm that they are not installed and not enabled. Which, again, is also not in sync with the project.yaml that says they ARE installed.

When running the command to enable them Craft complains that it doesn't recognise the handle, which makes no sense, because I use the right handle, which is proven by the fact that they show exactly like that in the plugins list.

"Failed to install plugin-handle: No plugin exists with the handle "plugin-handle" "

There is the issue.

**Reply by bencresty** [reply]
Alright, have it working now.
I needed to REMOVE my local plugins with composer, than UNINSTALL WITH FORCE the local plugins with Craft, than REQUIRE/INSTALL my local plugins again with composer and then INSTALL the local plugins with Craft.

Again; this is a clear indicator that two sources of truth is prone to errors. Situations like this would never occur when just having a single source of truth. Also during all this we encountered quite some other issues that didn't help debugging this:
- Craft shows 'successful', while it was not succesful, so that is a counter productive message
- Craft couldn't find plugin handles (at least that's what it told me), while the handles were perfectly fine known and never changed
- Crafts update/sync system gets messedup when things are out of sync. It is complaining when there are already database fields and when we remove the table it is complaining that there's no table. So either way we're quite fucked when things like this happen.
- The system of keeping project.yaml and database in sync often just doesn't work. All plugins were installed here just fine and the project.yaml said they were enabled, but still the system didn't work, the cms showed that these were 'disabled', which clearly was not the case, and nothing in the cms detected that the project.yaml and database weren't in sync and should be synced.

Again: the system shouldn't be out of sync in the first place. There should not even be a sync. There should be 1 single source of truth to have a reliable system. This is asking for trouble and caused already many headaches and issues here last years. These now were yet other issues being proof of that this system doesn't work and is very critical and prone to errors. And when out of sync things even get worse, because the system is not fixing things, is not capable of restoring things and is preventing us from getting the right messages to even know where to look. Yet another days wasted on a system that was just working fine before.
But it doesn't matter how many issues happen, you won't change your min whatever I say. So I'll leave it here. 

I appreciate your help now so thanks for that. But honestly, truth is that these issues shouldn't have occurred in the first place and wouldn't if there was only one single source of truth.
